entropy/coin_flip/app/src/flip_coin.ts

```diff
@@ -29,75 +45,80 @@ const parser = yargs(hideBin(process.argv))
+        sequenceNumber = log.args.sequenceNumber as bigint;
+      }),
   });
 
-  const web3 = new Web3(provider as any);
+ await client.watchBlocks({
+    onBlock: (block) => {
+      console.log(`\nPolling Block: ${block.number}`);
+      coinFlipContract.watchEvent.FlipResult({
+        onLogs: (logs) =>
+          logs.forEach((log) => {
+            if (log.args.sequenceNumber === sequenceNumber) {
+              console.log(
+                `\nFlip Result: ${log.args.isHeads ? "Heads" : "Tails"}`
+              );
+              process.exit(0);
+            }
+          }),
+      });
+    },
+  });
 
-  const coinFlipContract = new web3.eth.Contract(
-    CoinFlipAbi as any,
-    coinFlipContractAddress
+  const flipTxHash = await coinFlipContract.write.requestFlip(
+    [randomNumber as `0x${string}`],
+    { value: flipFee }
   );
+  console.log(`\nTransaction Hash: ${flipTxHash}`);
 
-  console.log(`Running coin flip prototcol.`);
+  const receipt = await client.waitForTransactionReceipt({
+    hash: flipTxHash,
+  });
 
-  console.log("1. Generating user's random number...");
-  const randomNumber = web3.utils.randomHex(32);
-  console.log(`   number    : ${randomNumber}`);
-
-  console.log("2. Requesting coin flip...");
-  const flipFee = await coinFlipContract.methods.getFlipFee().call();
-  console.log(`   fee       : ${flipFee} wei`);
-
-  const receipt = await coinFlipContract.methods
-    .requestFlip(randomNumber)
-    .send({ value: flipFee, from: provider.getAddress(0) });
-
-  console.log(`   tx        : ${receipt.transactionHash}`);
-  const sequenceNumber = receipt.events.FlipRequest.returnValues.sequenceNumber;
-  console.log(`   sequence  : ${sequenceNumber}`);
-
-  console.log("3. Waiting for result...");
-  // Poll for new FlipResult events emitted by the CoinFlip contract. It checks if the event
-  // has the same sequenceNumber as the request. If it does,
-  // it logs the result and stops polling.
-  let fromBlock = receipt.blockNumber;
-  const intervalId = setInterval(async () => {
-    const currentBlock = await web3.eth.getBlockNumber();
-
-    if (fromBlock > currentBlock) {
-      return;
-    }
-
-    // Get 'FlipResult' events emitted by the CoinFlip contract for given block range.
-    const events = await coinFlipContract.getPastEvents("FlipResult", {
-      fromBlock: fromBlock,
-      toBlock: currentBlock,
-    });
-    fromBlock = currentBlock + 1;
-
-    // Find the event with the same sequence number as the request.
-    const event = events.find(
-      (event) => event.returnValues.sequenceNumber === sequenceNumber
-    );
-
-    // If the event is found, log the result and stop polling.
-    if (event !== undefined) {
-      console.log(
-        `   result    : ${event.returnValues.isHeads ? "Heads" : "Tails"}`
-      );
-      clearInterval(intervalId);
-    }
-  }, 1000);
-
-  provider.engine.stop();
+  let receiptBlockNumber = receipt.blockNumber;
```

Comment:
Generally you don't want to introduce unnecessary bindings like this (just use `receipt.blockNumber` below) and if you do introduce it as a shorthand, you should use `const` and not `let`.

General advice is to nearly never use `let` and if you have to, really think about if the code you're writing has to be written mutably